packages/std/src/coins.rs

```diff
@@ -0,0 +1,346 @@
+            .filter(|coin| !coin.amount.is_zero())
+            .map(|coin| (coin.denom, coin.amount))
+            .collect::<BTreeMap<_, _>>();
+
+        // the map having a different length from the vec means the vec must either
+        // 1) contain duplicate denoms, or 2) contain zero amounts
+        if map.len() != vec_len {
+            return Err(StdError::parse_err(
+                type_name::<Self>(),
+                "duplicate denoms or zero amount",
+            ));
+        }
+
+        Ok(Self(map))
+    }
+}
+
+impl TryFrom<&[Coin]> for Coins {
+    type Error = StdError;
+
+    fn try_from(slice: &[Coin]) -> StdResult<Self> {
+        slice.to_vec().try_into()
+    }
+}
+
+impl<const N: usize> TryFrom<[Coin; N]> for Coins {
+    type Error = StdError;
+
+    fn try_from(slice: [Coin; N]) -> StdResult<Self> {
+        slice.to_vec().try_into()
+    }
+}
+
+impl TryFrom<Coin> for Coins {
+    type Error = StdError;
+
+    fn try_from(coin: Coin) -> StdResult<Self> {
+        let mut coins = Coins::default();
+        coins.add(coin)?;
+        Ok(coins)
+    }
+}
+
+impl From<Coins> for Vec<Coin> {
+    fn from(value: Coins) -> Self {
+        value.into_vec()
+    }
+}
+
+impl FromStr for Coins {
+    type Err = StdError;
+
+    fn from_str(s: &str) -> StdResult<Self> {
+        s.split(',')
+            .map(Coin::from_str)
+            .collect::<Result<Vec<_>, _>>()?
+            .try_into()
+    }
+}
+
+impl fmt::Display for Coins {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let s = self
+            .0
+            .iter()
+            .map(|(denom, amount)| format!("{amount}{denom}"))
+            .collect::<Vec<_>>()
+            .join(",");
+        write!(f, "{s}")
+    }
+}
+
+impl PartialEq<Coin> for Coins {
+    fn eq(&self, other: &Coin) -> bool {
+        self.0.len() == 1 && self.amount_of(&other.denom) == other.amount
+    }
+}
+
+impl Coins {
+    /// Cast to Vec<Coin>, while NOT consuming the original object
+    pub fn to_vec(&self) -> Vec<Coin> {
+        self.0
+            .iter()
+            .map(|(denom, amount)| Coin {
+                denom: denom.clone(),
+                amount: *amount,
+            })
+            .collect()
+    }
+
+    /// Cast to Vec<Coin>, consuming the original object
+    pub fn into_vec(self) -> Vec<Coin> {
+        self.0
+            .into_iter()
+            .map(|(denom, amount)| Coin { denom, amount })
+            .collect()
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Returns the denoms as a vector of strings.
+    /// The vector is guaranteed to not contain duplicates and sorted alphabetically.
+    pub fn denoms(&self) -> Vec<String> {
+        self.0.keys().cloned().collect()
+    }
+
+    /// Returns the amount of the given denom or zero if the denom is not present.
+    pub fn amount_of(&self, denom: &str) -> Uint128 {
```

Comment:
What about making this an Option<Uint128>? The user can do unwrap_or_default on their own.

By the way, are zero amounts allowed inside of Coins?

Reply:
Zero amounts should not be allowed (unless I overlooked some corner case somewhere), so IMHO the Option doesn't really add much value here, but I'm also fine with adding it for consistency